common/src/main/java/io/netty/util/internal/StringUtil.java

```diff
@@ -260,67 +260,69 @@ public static CharSequence escapeCsv(CharSequence value) {
+        if (start > last) {
+            return EMPTY_STRING;
+        }
+
+        int firstUnescapedSpecial = -1;
+        boolean quoted = false;
+        if (isDoubleQuote(value.charAt(start))) {
+            quoted = isDoubleQuote(value.charAt(last)) && last > start;
+            if (quoted) {
+                start++;
+                last--;
+            } else {
+                firstUnescapedSpecial = start;
             }
         }
 
-        StringBuilder result = new StringBuilder(length + CSV_NUMBER_ESCAPE_CHARACTERS);
-        boolean quoted = isDoubleQuote(value.charAt(start)) && isDoubleQuote(value.charAt(last)) && length != 1;
-        boolean foundSpecialCharacter = false;
-        boolean escapedDoubleQuote = false;
-        for (int i = start; i <= last; i++) {
-            char current = value.charAt(i);
-            switch (current) {
-                case DOUBLE_QUOTE:
-                    if (i == start || i == last) {
-                        if (!quoted) {
-                            result.append(DOUBLE_QUOTE);
-                        } else {
-                            continue;
-                        }
-                    } else {
-                        boolean isNextCharDoubleQuote = isDoubleQuote(value.charAt(i + 1));
-                        if (!isDoubleQuote(value.charAt(i - 1)) &&
-                                (!isNextCharDoubleQuote || i + 1 == last)) {
-                            result.append(DOUBLE_QUOTE);
-                            escapedDoubleQuote = true;
-                        }
+        if (firstUnescapedSpecial < 0) {
+            for (int i = start; i <= last; i++) {
+                char c = value.charAt(i);
+                if (!quoted && (c == LINE_FEED || c == CARRIAGE_RETURN || c == COMMA)) {
+                    firstUnescapedSpecial = i;
+                    break;
+                }
+                if (isDoubleQuote(c)) {
+                    if (i == last || !isDoubleQuote(value.charAt(i + 1))) {
+                        firstUnescapedSpecial = i;
                         break;
                     }
-                case LINE_FEED:
-                case CARRIAGE_RETURN:
-                case COMMA:
-                    foundSpecialCharacter = true;
+                    i++;
+                }
             }
-            result.append(current);
-        }
 
-        if (escapedDoubleQuote || foundSpecialCharacter && !quoted) {
-            return quote(result);
-        }
-        if (trimmed) {
-            return quoted ? quote(result) : result;
+            if (firstUnescapedSpecial < 0) {
+                // Special characters is not found or all of them already escaped.
+                // In the most cases returns a same string. New string will be instantiated (via StringBuilder)
+                // only if it really needed. It's important to prevent GC extra load.
+                return value.subSequence(quoted? start - 1 : start, last + (quoted? 2 : 1));
```

Comment:
have you considered evaluating the `quoted` once here? `return quoted ? value.subSequence(...) : value.subSequence(...);`